### PR TITLE
E1713. Refactor penalty_helper.rb and late_policies_controller.rb

### PR DESCRIPTION
**Pull request by michaelamoran**

### CODE REVIEW

**Comment by michaelamoran** [comment]
I will be modifying our code this weekend to fix the issues.

**Comment by Winbobob** [comment]
Hi team, do you finish modifying the code?

**Comment by michaelamoran** [comment]
I'm going to need some guidance on the remaining items. Many of them cannot be avoided or should not be avoided. Some of them I simply don't understand. Some of them I have changed in an attempt to appease it but it doesn't like any of the possibilities for one reason or another.

File: app/helpers/penalty_helper.rb [hunk 1]
```diff
@@ -1,5 +1,6 @@
 module PenaltyHelper
-  def calculate_penalty(participant_id)
+
+  def self.calculate_penalty(participant_id)
```
Comment (Winbobob) [inline app/helpers/penalty_helper.rb]: I do not think it is necessary to make this method a module class method.

File: app/helpers/penalty_helper.rb [hunk 2]
```diff
@@ -34,28 +33,17 @@ def calculate_penalty(participant_id)
     penalties
   end
 
-  def set_penalty_policy
-    @late_policy = Assignment.find(@assignment_id).late_policy
-    @late_policy.max_penalty
-  end
-
-  def calculate_submission_penalty
+  def self.calculate_submission_penalty
```
Comment (Winbobob) [inline app/helpers/penalty_helper.rb]: I do not think it is necessary to make this method a module class method.

File: app/helpers/penalty_helper.rb [hunk 3]
```diff
@@ -67,7 +55,7 @@ def calculate_submission_penalty
     end
   end
 
-  def calculate_review_penalty
+  def self.calculate_review_penalty
```
Comment (Winbobob) [inline app/helpers/penalty_helper.rb]: I do not think it is necessary to make this method a module class method.

File: app/helpers/penalty_helper.rb [hunk 4]
```diff
@@ -84,7 +72,7 @@ def calculate_review_penalty
     penalty
   end
 
-  def calculate_meta_review_penalty
+  def self.calculate_meta_review_penalty
```
Comment (Winbobob) [inline app/helpers/penalty_helper.rb]: I do not think it is necessary to make this method a module class method.

File: app/helpers/penalty_helper.rb [hunk 5]
```diff
@@ -100,51 +88,9 @@ def calculate_meta_review_penalty
...
-    penalty
-  end
-
-  def calculate_team_feedback_penalty
-    penalty = 0
-    penalty
-  end
-
-  def compute_penalty_on_reviews(review_mappings, review_due_date, num_of_reviews_required)
+  def self.compute_penalty_on_reviews(review_mappings, review_due_date, num_of_reviews_required)
```
Comment (Winbobob) [inline app/helpers/penalty_helper.rb]: I do not think it is necessary to make this method a module class method.

File: app/helpers/penalty_helper.rb [hunk 6]
```diff
@@ -181,4 +122,57 @@ def compute_penalty_on_reviews(review_mappings, review_due_date, num_of_reviews_
     end
     penalty
   end
+
+  def self.calculate_penalty_units(time_difference, penalty_unit)
```
Comment (Winbobob) [inline app/helpers/penalty_helper.rb]: I do not think it is necessary to make this method a module class method.

File: app/helpers/penalty_helper.rb [hunk 7]
```diff
@@ -181,4 +122,57 @@ def compute_penalty_on_reviews(review_mappings, review_due_date, num_of_reviews_
...
+      penalty_units = time_difference / 60
+    elsif penalty_unit == 'Hour'
+      penalty_units = time_difference / 3600
+    elsif penalty_unit == 'Day'
+      penalty_units = time_difference / 86_400
+    end
+  end
+
+  # checking that penalty_per_unit is not exceeding max_penalty
+  def self.check_penalty_points_validity(max_penalty, penalty_per_unit)
```
Comment (Winbobob) [inline app/helpers/penalty_helper.rb]: I do not think it is necessary to make this method a module class method.

File: app/helpers/penalty_helper.rb [hunk 8]
```diff
@@ -181,4 +122,57 @@ def compute_penalty_on_reviews(review_mappings, review_due_date, num_of_reviews_
...
+    return max_penalty < penalty_per_unit
+  end
+
+  # method to check whether the policy name given as a parameter already exists under the current instructor id
+  # it return true if there's another policy with the same name under current instructor else false
+  def self.check_policy_with_same_name(late_policy_name, instructor_id)
```
Comment (Winbobob) [inline app/helpers/penalty_helper.rb]: I do not think it is necessary to make this method a module class method.

File: app/helpers/penalty_helper.rb [hunk 9]
```diff
@@ -181,4 +122,57 @@ def compute_penalty_on_reviews(review_mappings, review_due_date, num_of_reviews_
...
+        next unless p.instructor_id == instructor_id
+        return true
+      end
+    end
+    return false
+  end
+
+  # this method updates all the penalty objects which uses the penalty policy which is passed as a parameter
+  # whenever a policy is updated, all the existing penalty objects needs to be updated according to new policy
+  def self.update_calculated_penalty_objects(penalty_policy)
```
Comment (Winbobob) [inline app/helpers/penalty_helper.rb]: I do not think it is necessary to make this method a module class method.

File: app/helpers/penalty_helper.rb (hunk 7 above)
Comment (Winbobob) [inline app/helpers/penalty_helper.rb]: And this method is unnecessary.

File: spec/helpers/penalty_helper_spec.rb [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+    it "should return 1 when passed a 1 day time difference and a penalty unit of 'Day'" do
+      penalty_unit = 'Day'
+      time_difference = DateTime.now.in_time_zone - (DateTime.now.in_time_zone - 1.day)
+      penalty_minutes = PenaltyHelper.calculate_penalty_units(time_difference, penalty_unit)
+      expect(penalty_minutes.round).to be == 1
+    end
+  end
+
+  it "has a valid factory" do
```
Comment (Winbobob) [inline spec/helpers/penalty_helper_spec.rb]: This test case is useless!

File: spec/helpers/penalty_helper_spec.rb [hunk 11]
```diff
@@ -0,0 +1,72 @@
+require 'rails_helper'
+
+describe "PenaltyHelper" do
+  before(:each) do
+    @course_participant = create(:course_participant)
```
Comment (Winbobob) [inline spec/helpers/penalty_helper_spec.rb]: Use `create` un helper test is overkill. Use `build` instead.

**Comment by Winbobob** [comment]
Methods from `app/controllers/late_policies_controller.rb` should be placed into `app/models/late_policy.rb`, instead of `app/helpers/penalty_helper.rb`.

**Pull request merged**